<a href="https://colab.research.google.com/github/ZeyadSabbah/TrivagoRecommenderSystem/blob/master/TrivagoFeatureEngineering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Feature Engineering

## Mounting to Drive

In [0]:
from google.colab import drive
drive.mount('/content/drive')

In [0]:
%cd /content/drive/My Drive/Trivago/Project/TrivagoRecommenderSystem

## Loading Libraries & Datasets

In [0]:
import pandas as pd
import numpy as np
from datetime import datetime
from datetime import timedelta
import math
import matplotlib.pyplot as plt
from datetime import datetime
import re
import random
from sklearn.preprocessing import OneHotEncoder
from sklearn.metrics.pairwise import cosine_similarity
from collections import Counter
from sklearn.preprocessing import MultiLabelBinarizer
from sklearn.decomposition import PCA

In [0]:
item_metadata_filepath = './Datasets/raw_data/item_metadata.csv'
item_metadata = pd.read_csv(item_metadata_filepath)

train_filepath = './Datasets/raw_data/Sets/train.csv'
train = pd.read_csv(train_filepath)

In [5]:
train.head()

user_id     session_id   timestamp  ...  current_filters impressions prices
0  00RL8Z82B2Z1  aff3928535f48  1541037460  ...              NaN         NaN    NaN
1  00RL8Z82B2Z1  aff3928535f48  1541037522  ...              NaN         NaN    NaN
2  00RL8Z82B2Z1  aff3928535f48  1541037522  ...              NaN         NaN    NaN
3  00RL8Z82B2Z1  aff3928535f48  1541037532  ...              NaN         NaN    NaN
4  00RL8Z82B2Z1  aff3928535f48  1541037532  ...              NaN         NaN    NaN

[5 rows x 12 columns]

## Important Dataframes Preparation

In engineer global features from the train dataset, which will be most of the cases around either clickouts or final click out examples (instances), two dataframes of them will be created.

In order to get representative global values to the items, duplications must be removed. Rows would not all be duplicated because of some attributes such as timestamp, reference, and step.

In [0]:
ClickoutDF = train[train.action_type=='clickout item']
FinalClickoutDF = train[train.action_type=='clickout item'].groupby('session_id').tail(1)
ClickoutUniqueDF = ClickoutDF.drop_duplicates(subset=['session_id', 'impressions'], keep='first')

## Item Global Features
### Number of Properties

In [7]:
item_metadata.properties = item_metadata.properties.apply(lambda x: x.split('|'))
item_metadata['NumberOfProperties'] = item_metadata.properties.apply(lambda x: len(x))
item_metadata.head(2)

item_id  ... NumberOfProperties
0     5101  ...                 62
1     5416  ...                 46

[2 rows x 3 columns]

In [8]:
#getting total number of unique properties across all items
AllPropertiesList = item_metadata.properties.tolist()

AllPropertiesFlatList = []
for sublist in AllPropertiesList:
    for item in sublist:
        AllPropertiesFlatList.append(item)
        
print('Number of unique properties is', len(set(AllPropertiesFlatList)))

Number of unique properties is 157


### Items Properties Similarities
The purpose is to get cosine similarity between items. The maximum 25 items out of the list can be extracted as a dataframe and cosine similarity can be obtained in order to get the similar items to the ones the user had interacted with to be listed on the top of the list.  

In [9]:
item_metadata.properties = item_metadata.properties.apply(lambda x: tuple(x))

one_hot = MultiLabelBinarizer()
properties_encoded = one_hot.fit_transform(item_metadata.properties.values.tolist())
properties_encodedDF = pd.DataFrame(properties_encoded)

#changing column names
properties_list = one_hot.classes_.tolist()
for i in range(len(properties_list)):
    properties_encodedDF = properties_encodedDF.rename(columns={i:properties_list[i]})

#creating a column of the item id to get the similarity between items
item_metadata.item_id = item_metadata.item_id.apply(lambda x: str(x))
properties_encodedDF['item_id'] = item_metadata.item_id

properties_encodedDF.head()

1 Star  2 Star  3 Star  ...  WiFi (Public Areas)  WiFi (Rooms)  item_id
0       0       0       0  ...                    1             1     5101
1       0       0       0  ...                    1             1     5416
2       0       0       1  ...                    1             1     5834
3       0       0       0  ...                    1             0     5910
4       0       0       0  ...                    1             1     6066

[5 rows x 158 columns]

#### K-means

In [13]:
from sklearn.cluster import KMeans

X = properties_encodedDF.drop(columns='item_id')
X = X.values
kmeans = KMeans(n_clusters=5, random_state=0).fit(X)
kmeans.labels_

array([0, 0, 0, ..., 1, 1, 2], dtype=int32)

In [0]:
PropertiesClustersDF = pd.DataFrame({'item_id':properties_encodedDF.item_id.values.tolist(),
                                     'cluster':kmeans.labels_.tolist()})
PropertiesClustersDF.cluster = PropertiesClustersDF.cluster.replace([0, 1, 2, 3, 4], ['A', 'B', 'C', 'D', 'E'])

In [15]:
ClusterDF = pd.get_dummies(PropertiesClustersDF.cluster)
ClusterDF['item'] = PropertiesClustersDF.item_id.values.tolist()
ClusterDF.head()

A  B  C  D  E  item
0  1  0  0  0  0  5101
1  1  0  0  0  0  5416
2  1  0  0  0  0  5834
3  1  0  0  0  0  5910
4  1  0  0  0  0  6066

In [0]:
ClusterDF.to_csv('./Datasets/clean_data/ItemsPropertiesSimilarities.csv', index=False)

#### PCA

In [18]:
pca = PCA(n_components=5)
X = properties_encodedDF.drop(columns='item_id')
properties = pca.fit_transform(X)
propertiesDF = pd.DataFrame(data = properties, columns=['A', 'B', 'C', 'D', 'E'])
propertiesDF['item_id'] = properties_encodedDF.item_id.values.tolist()

propertiesDF.head()

A         B         C         D         E item_id
0 -2.622219 -0.902672 -0.723547  0.297539 -0.097791    5101
1 -2.526538 -0.790910 -0.666546  0.287467  0.253292    5416
2 -2.214742 -0.727124 -0.685831  0.305293  0.629916    5834
3 -2.115618 -0.536262  0.169221  0.105751  0.301906    5910
4 -2.463044 -0.854214 -0.718518  0.179768  0.615274    6066

In [0]:
propertiesDF.to_csv('./Datasets/clean_data/ItemsPropertiesSimilarities/PCA_Properties.csv')

### Number of Times in Impressions
The purpose of this feature is to check how many times an item had been shown to users in the list.

Finding number of times an item has been mentioned in a list.

In [0]:
AllImpressionsList = ClickoutUniqueDF.impressions.apply(lambda x:x.split('|'))

AllImpressionsFlatList = []
for sublist in AllImpressionsList:
    for item in sublist:
        AllImpressionsFlatList.append(item)

InImpressionsCounter = Counter(AllImpressionsFlatList)
InImpressionsDF = pd.DataFrame.from_dict(InImpressionsCounter, orient='index').reset_index()\
                              .rename(columns={'index':'item_id', 0:'NumberInImpressions'})
InImpressionsDF.head(2)

item_id  NumberInImpressions
0  3400638                  394
1  1253714                   74

In [0]:
InImpressionsDF.item_id.nunique(), len(InImpressionsDF), item_metadata.item_id.nunique()

(777243, 777243, 927142)

Number of items in this dataframe is less than the number of items in item_metadata, that's because some of the items had not been mentioned in the impressions list.

In [0]:
item_metadata.item_id = item_metadata.item_id.apply(lambda x: str(x))
item_metadata.drop(columns='properties', inplace=True)

#left joining
item_metadata = item_metadata.merge(InImpressionsDF, on='item_id', how='left')
item_metadata.head(2)

item_id  NumberOfProperties  NumberInImpressions
0    5101                  62                 68.0
1    5416                  46                 58.0

### Number of Times in Reference
The purpose of this feature is to check how many times an item has been mentioned in the Reference attribute in the whole train set. 

In [0]:
InReferencesCounter = Counter(train.reference.values.tolist())
InReferencesDF = pd.DataFrame.from_dict(InReferencesCounter, orient='index').reset_index()\
                              .rename(columns={'index':'item_id', 0:'NumberInReferences'})

InReferencesDF.head(2)

item_id  NumberInReferences
0  Newtown                  13
1   666856                  23

In [0]:
#left joining
item_metadata = item_metadata.merge(InReferencesDF, on='item_id', how='left')
item_metadata.head(2)

item_id  NumberOfProperties  NumberInImpressions  NumberInReferences
0    5101                  62                 68.0                 9.0
1    5416                  46                 58.0                41.0

### Number of Times in Clickout
The purpose of this feature is to check how many times an item has been clicked out

In [0]:
InClickoutCounter = Counter(ClickoutDF.reference.values.tolist())
InClickoutDF = pd.DataFrame.from_dict(InClickoutCounter, orient='index').reset_index()\
                              .rename(columns={'index':'item_id', 0:'NumberAsClickout'})
InClickoutDF.head(2)

item_id  NumberAsClickout
0   109038                45
1  1257342                19

In [0]:
#left joining
item_metadata = item_metadata.merge(InClickoutDF, on='item_id', how='left')
item_metadata.head(2)

item_id  NumberOfProperties  ...  NumberInReferences  NumberAsClickout
0    5101                  62  ...                 9.0               4.0
1    5416                  46  ...                41.0               6.0

[2 rows x 5 columns]

### Number of Time in Final Clickout
The purpose of this feature is to get the number of times an item had been mentioned as a final clickout.

In [0]:
InFinalClickoutCounter = Counter(FinalClickoutDF.reference.values.tolist())
InFinalClickoutDF = pd.DataFrame.from_dict(InFinalClickoutCounter, orient='index').reset_index()\
                              .rename(columns={'index':'item_id', 0:'NumberAsFinalClickout'})
InFinalClickoutDF.head(2)

item_id  NumberAsFinalClickout
0  1257342                      6
1  2795374                     15

In [0]:
#left joining
item_metadata = item_metadata.merge(InFinalClickoutDF, on='item_id', how='left')
item_metadata.head(2)

item_id  NumberOfProperties  ...  NumberInReferences  NumberAsFinalClickout
0    5101                  62  ...                 9.0                    2.0
1    5416                  46  ...                41.0                    2.0

[2 rows x 5 columns]

The next step would be dividing the NumberAsFinalClickout by the other 3, in order to get FinalClickout Relativity

### Final Clickout To Impressions
The purpose of this feature is to get item's rate of clicking out when listed to the user.

In [0]:
FClickoutToImpressions = item_metadata.NumberAsFinalClickout/item_metadata.NumberInImpressions
FClickoutToImpressions.head(2)

0    0.029412
1    0.034483
dtype: float64

In [0]:
#adding attribute
item_metadata['FClickoutToImpressions'] = FClickoutToImpressions
item_metadata.head(2)

item_id  NumberOfProperties  ...  NumberAsFinalClickout  FClickoutToImpressions
0    5101                  62  ...                    2.0                0.029412
1    5416                  46  ...                    2.0                0.034483

[2 rows x 6 columns]

### Final Clickout To References
The purpose of this feature is to get item's rate of clicking out when it was interacted with.

In [0]:
FClickoutToReferences = item_metadata.NumberAsFinalClickout/item_metadata.NumberInReferences
FClickoutToReferences.head(2)

0    0.222222
1    0.048780
dtype: float64

In [0]:
#adding attribute
item_metadata['FClickoutToReferences'] = FClickoutToReferences
item_metadata.head(2)

item_id  NumberOfProperties  ...  FClickoutToImpressions  FClickoutToReferences
0    5101                  62  ...                0.029412               0.222222
1    5416                  46  ...                0.034483               0.048780

[2 rows x 7 columns]

### Final Clickout To Clickout
The purpose of this feature is to get item's rate of clickout when it was clicked out before.

In [0]:
FClickoutToClickout = item_metadata.NumberAsFinalClickout/item_metadata.NumberAsClickout
FClickoutToClickout.head(2)

In [0]:
#adding attribute
item_metadata['FClickoutToClickout'] = FClickoutToClickout
item_metadata.head(2)

### Item's Average Rank
The purpose of this feature is to get item's position in the list provided to the user across the train set.  
Since the purpose is to get the average rank across the lists shown to users, an important notice which is that some sessions have different reference and click outs, which provides the same impression list. Duplicated impressions lists in each session should be dropped. (Same thing applies to price as well.)

In [0]:
# using All Clickout dataframe, but the one with the unique impressions for each session for this feature
ClickoutUniqueDF.head(1)

In [0]:
SessionImpressionsDF = ClickoutUniqueDF[['session_id', 'impressions']]
SessionImpressionsDF.impressions = SessionImpressionsDF.impressions.apply(lambda x: x.split('|'))
SessionImpressionsDF.impressions.head(2)

In [0]:
SessionImpressionsDF_exploded = SessionImpressionsDF.explode('impressions')
SessionImpressionsDF_exploded = SessionImpressionsDF_exploded.rename(columns={'impressions':'item_id'})
SessionImpressionsDF_exploded.head(2)

In [0]:
#creating a rank column
rank = SessionImpressionsDF.impressions.apply(lambda x: list(range(1, (len(x) + 1))))
SessionRankDF = pd.concat([SessionImpressionsDF.session_id, rank], axis=1)
SessionRankDF = SessionRankDF.rename(columns={'impressions':'rank'})
SessionRankDF_exploded = SessionRankDF.explode('rank')
SessionRankDF_exploded.head(2)

In [0]:
#creating dataframe that combines both items and rank
ItemRankDF = pd.DataFrame({'item_id':SessionImpressionsDF_exploded['item_id'].values.tolist(),
                            'rank':SessionRankDF_exploded['rank'].values.tolist()})
ItemRankDF.head(2)

In [0]:
#getting the mean of ranks for each item
ItemAverageRank = ItemRankDF.groupby('item_id', sort=False)['rank'].mean().to_frame().reset_index()
ItemAverageRank = ItemAverageRank.rename(columns={'rank':'MeanRank'})
ItemAverageRank.head(2)

In [0]:
#left joining
item_metadata = item_metadata.merge(ItemAverageRank, on='item_id', how='left')
item_metadata.head(2)

### Item's Average Price
The purpose of this feature is to get item's average price accross the train set.  
Making sure that the length of each impressions list is as the same length of the prices list.

In [0]:
ImpressionsLength = ClickoutUniqueDF.impressions.apply(lambda x: x.split('|')).apply(lambda x: len(x))
PricesLength = ClickoutUniqueDF.prices.apply(lambda x: x.split('|')).apply(lambda x: len(x))
ImpressionsLength.equals(PricesLength)

In [0]:
SessionPricesDF = ClickoutUniqueDF[['session_id', 'prices']]
SessionPricesDF.prices = SessionPricesDF.prices.apply(lambda x: x.split('|'))
SessionPricesDF_exploded = SessionPricesDF.explode('prices')
ItemPriceDF = pd.DataFrame({'item_id':SessionImpressionsDF_exploded.item_id.values.tolist(),
                            'price':SessionPricesDF_exploded.prices.values.tolist()})
ItemPriceDF.head(2)

In [0]:
ItemPriceDF.price = ItemPriceDF.price.apply(lambda x: int(x))
ItemAveragePriceDF = ItemPriceDF.groupby('item_id', sort=False)['price'].mean().to_frame().reset_index()
ItemAveragePriceDF = ItemAveragePriceDF.rename(columns={'price':'MeanPrice'})
ItemAveragePriceDF.head(2)

In [0]:
item_metadata = item_metadata.merge(ItemAveragePriceDF, on='item_id', how='left')
item_metadata.head(2)

item_id  NumberOfProperties  ...   MeanRank   MeanPrice
0    5101                  62  ...  13.235294  121.617647
1    5416                  46  ...  10.431034  102.913793

[2 rows x 11 columns]

### Item's Maximum Price

In [0]:
ItemMaximumPriceDF = ItemPriceDF.groupby('item_id', sort=False)['price'].max().to_frame().reset_index()
ItemMaximumPriceDF = ItemMaximumPriceDF.rename(columns={'price':'MaxPrice'})
ItemMaximumPriceDF.head(2)

item_id  MaxPrice
0  3400638      2523
1  1253714       248

In [0]:
#left joining
item_metadata = item_metadata.merge(ItemMaximumPriceDF, on='item_id', how='left')
item_metadata.head(2)

item_id  NumberOfProperties  ...   MeanPrice  MaxPrice
0    5101                  62  ...  121.617647     241.0
1    5416                  46  ...  102.913793     270.0

[2 rows x 12 columns]

### Item's Minimum Price

In [0]:
ItemMinimumPriceDF = ItemPriceDF.groupby('item_id', sort=False)['price'].min().to_frame().reset_index()
ItemMinimumPriceDF = ItemMinimumPriceDF.rename(columns={'price':'MinPrice'})
ItemMinimumPriceDF.head(2)

item_id  MinPrice
0  3400638        68
1  1253714        50

In [0]:
#left joining
item_metadata = item_metadata.merge(ItemMinimumPriceDF, on='item_id', how='left')
item_metadata.head(2)

item_id  NumberOfProperties  ...  MaxPrice  MinPrice
0    5101                  62  ...     241.0      64.0
1    5416                  46  ...     270.0      59.0

[2 rows x 13 columns]

### Price Rank
The purpose of this feature is to get the average price rank of an item across the train set. Sorting the items not by what is being showed to the user, but by the items' prices, getting the average price rank for each item across the whole train. It shows where an item stands from its peers.  
By using the argsort function, the price rank can be feature can be engineered.

In [0]:
ClickoutDF.prices[13]

'95|66|501|112|95|100|101|72|82|56|56|143|70|25|71|162|73|143|188|118|77|131|143|49|165'

In [0]:
np.argsort(list(map(int, ClickoutDF.prices[13].split('|'))))

array([13, 23,  9, 10,  1, 12, 14,  7, 16, 20,  8,  0,  4,  5,  6,  3, 19,
       21, 17, 22, 11, 15, 24, 18,  2])

In [0]:
SessionPricesDF['PriceRank'] = SessionPricesDF.prices.apply(lambda x: (list(map(int, x))))\
                                                     .apply(lambda x: np.argsort(x))
SessionPricesDF.head(2)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


session_id  ...                                          PriceRank
13  aff3928535f48  ...  [13, 23, 9, 10, 1, 12, 14, 7, 16, 20, 8, 0, 4,...
15  aff3928535f48  ...  [8, 1, 15, 5, 12, 16, 20, 9, 4, 10, 23, 24, 6,...

[2 rows x 3 columns]

In [0]:
SessionPricesRankDF = SessionPricesDF[['session_id', 'PriceRank']].explode('PriceRank')
SessionPricesRankDF.head(2)

session_id PriceRank
13  aff3928535f48        13
13  aff3928535f48        23

In [0]:
ItemPriceRankDF = pd.DataFrame({'item_id':SessionImpressionsDF_exploded.item_id.values.tolist(),
                            'PriceRank':SessionPricesRankDF.PriceRank.values.tolist()})
ItemPriceRankDF.head(2)

item_id  PriceRank
0  3400638         13
1  1253714         23

In [0]:
#getting the average price rank of each item
ItemAveragePriceRankDF = ItemPriceRankDF.groupby('item_id', sort=False).PriceRank.mean().to_frame().reset_index()
ItemAveragePriceRankDF = ItemAveragePriceRankDF.rename(columns={'PriceRank':'AveragePriceRank'})
ItemAveragePriceRankDF.head(2)

item_id  AveragePriceRank
0  3400638         11.901015
1  1253714         10.770270

In [0]:
# left joining
item_metadata = item_metadata.merge(ItemAveragePriceRankDF, on='item_id', how='left')
item_metadata.head(2)

item_id  NumberOfProperties  ...  MinPrice  AveragePriceRank
0    5101                  62  ...      64.0         12.985294
1    5416                  46  ...      59.0         11.327586

[2 rows x 14 columns]

In [0]:
#filling NaN values with zeros
item_metadata = item_metadata.fillna(0)

GlobalItemFeatures = item_metadata.merge(properties_encodedDF, on='item_id', how='left')
GlobalItemFeatures.head(2)

item_id  NumberOfProperties  ...  WiFi (Public Areas)  WiFi (Rooms)
0    5101                  62  ...                    1             1
1    5416                  46  ...                    1             1

[2 rows x 171 columns]

## Item Local Features

### General Features

In [0]:
session_gb = train.groupby('session_id', sort=False)
train.session_id.nunique()

610596

#### item_id

In [0]:
def get_data_clickout(data):
  data_clickout = data[data['action_type']=='clickout item'].groupby('session_id').tail(1)
  return data_clickout

def get_item_id(data_clickout):
  item_id = data_clickout[['session_id', 'impressions']]
  item_id['impressions'] = item_id['impressions'].apply(lambda x: x.split('|'))
  item_id = item_id.explode('impressions')
  item_id = item_id.rename(columns={'impressions':'item_id'})
  item_id = item_id.reset_index(drop=True)
  return item_id

In [0]:
FinalClickoutDF = get_data_clickout(train)
item_id = get_item_id(FinalClickoutDF)
item_id

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  import sys


session_id  item_id
0         aff3928535f48    55109
1         aff3928535f48   129343
2         aff3928535f48    54824
3         aff3928535f48  2297972
4         aff3928535f48   109014
...                 ...      ...
12586682  62728015bec05  2712342
12586683  62728015bec05    48497
12586684  62728015bec05    11933
12586685  62728015bec05  1714483
12586686  62728015bec05  1236687

[12586687 rows x 2 columns]

#### clickout

In [0]:
def get_clickout(data, data_clickout, item_id):
    clickout = data_clickout[['session_id','reference']]
    clickout = item_id.merge(clickout, on='session_id', how='left')
    clickout['clickout'] = clickout.apply(lambda x: 1 if x['item_id'] == x['reference'] else 0, axis=1)
    clickout.drop(columns='reference', inplace=True)
    clickout = clickout.reset_index(drop=True)
    IndexToNan = data_clickout.reference.index.values.tolist()
    data.loc[IndexToNan, 'reference'] = np.nan
    return clickout, data

In [0]:
clickout, train = get_clickout(train, FinalClickoutDF, item_id)
clickout.head(10)

session_id  item_id  clickout
0  aff3928535f48    55109         0
1  aff3928535f48   129343         0
2  aff3928535f48    54824         0
3  aff3928535f48  2297972         0
4  aff3928535f48   109014         0
5  aff3928535f48  1257342         1
6  aff3928535f48  1031578         0
7  aff3928535f48   109018         0
8  aff3928535f48  1332971         0
9  aff3928535f48   666916         0

#### price

In [0]:
def get_price(data_clickout):
  price = data_clickout[['session_id', 'prices']]
  price['prices'] = price['prices'].apply(lambda x: x.split('|'))
  price = price.explode('prices')
  price['prices'] = price['prices'].apply(lambda x: int(x))
  price = price.rename(columns={'prices':'price'})
  price = price.reset_index(drop=True)
  return price

In [0]:
price = get_price(FinalClickoutDF)
price

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


session_id  price
0         aff3928535f48    162
1         aff3928535f48     25
2         aff3928535f48    150
3         aff3928535f48    143
4         aff3928535f48    101
...                 ...    ...
12586682  62728015bec05     73
12586683  62728015bec05    169
12586684  62728015bec05     87
12586685  62728015bec05    485
12586686  62728015bec05    171

[12586687 rows x 2 columns]

#### item_rank

In [0]:
def get_item_rank(data_clickout):
  item_rank = data_clickout[['session_id', 'impressions']]
  item_rank['impressions'] = item_rank['impressions'].apply(lambda x: x.split('|'))
  item_rank['impressions'] = item_rank['impressions'].apply(lambda x: list(range(1, len(x) + 1)))
  item_rank = item_rank.explode('impressions')
  item_rank['impressions'] = item_rank['impressions'].apply(lambda x: int(x))
  item_rank = item_rank.rename(columns={'impressions':'item_rank'})
  item_rank = item_rank.reset_index(drop=True)
  return item_rank

In [0]:
item_rank = get_item_rank(FinalClickoutDF)
item_rank

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


session_id  item_rank
0         aff3928535f48          1
1         aff3928535f48          2
2         aff3928535f48          3
3         aff3928535f48          4
4         aff3928535f48          5
...                 ...        ...
12586682  62728015bec05         21
12586683  62728015bec05         22
12586684  62728015bec05         23
12586685  62728015bec05         24
12586686  62728015bec05         25

[12586687 rows x 2 columns]

#### price_rank

In [0]:
def get_price_rank(data):
  price_rank = data.groupby('session_id', sort=False).price.apply(lambda x: x.values).to_frame().reset_index().rename(columns={'price':'price_list'})
  price_rank.price_list = price_rank.price_list.apply(lambda x: np.argsort(x))
  price_rank = price_rank.rename(columns={'price_list':'price_rank'})
  price_rank = price_rank.explode('price_rank')
  price_rank = price_rank.reset_index(drop=True)
  return price_rank

In [0]:
price_rank = get_price_rank(price)
price_rank

session_id price_rank
0         aff3928535f48          8
1         aff3928535f48          1
2         aff3928535f48         15
3         aff3928535f48          5
4         aff3928535f48         12
...                 ...        ...
15373343  62728015bec05          6
15373344  62728015bec05         19
15373345  62728015bec05         10
15373346  62728015bec05         23
15373347  62728015bec05         16

[15373348 rows x 2 columns]

#### session_duration

In [0]:
def get_session_duration(data, item_id):
  session_duration = data.groupby('session_id', sort=False).timestamp.max() - data.groupby('session_id', sort=False).timestamp.min()
  session_duration = session_duration.to_frame().rename(columns={'timestamp':'session_duration'})
  session_duration = item_id.merge(session_duration, on='session_id', how='left')
  session_duration.drop(columns='item_id', inplace=True)
  session_duration = session_duration.reset_index(drop=True)
  return session_duration

In [0]:
session_duration = get_session_duration(train, item_id)
session_duration

session_id  session_duration
0         aff3928535f48              1025
1         aff3928535f48              1025
2         aff3928535f48              1025
3         aff3928535f48              1025
4         aff3928535f48              1025
...                 ...               ...
15373343  62728015bec05               571
15373344  62728015bec05               571
15373345  62728015bec05               571
15373346  62728015bec05               571
15373347  62728015bec05               571

[15373348 rows x 2 columns]

#### item_duration

In [0]:
def get_item_duration(data, item_id):
  item_duration = data.groupby(['session_id', 'reference'], sort=False).timestamp.max() - data.groupby(['session_id', 'reference'], sort=False).timestamp.min()
  item_duration = item_duration.reset_index().rename(columns={'reference':'item_id', 'timestamp':'item_duration'})
  item_duration = item_id.merge(item_duration, left_on=['session_id', 'item_id'], right_on=['session_id', 'item_id'], how='left')
  item_duration = item_duration.fillna(0)
  item_duration = item_duration.reset_index(drop=True)
  return item_duration

In [0]:
item_duration = get_item_duration(train, item_id)
item_duration[item_duration.item_duration>0]

session_id  item_id  item_duration
25        3599a6f709eab  2795374          134.0
65        ec139e10b9238  1032816          338.0
79        325fafb5fa450    65685           65.0
126       7157899be2839  2552514            3.0
185       4c6062d7cefe4   138930           10.0
...                 ...      ...            ...
15373250  92b7ab1287edf   207081           14.0
15373251  92b7ab1287edf    50584           82.0
15373273  6b66cb0cfb518  3811156            3.0
15373274  6b66cb0cfb518  2337580           10.0
15373323  62728015bec05  6617798           21.0

[530222 rows x 3 columns]

#### item_session_duration

In [0]:
def get_item_session_duration(item_duration, session_duration):
  item_duration['item_session_duration'] = item_duration.item_duration/session_duration.session_duration
  item_session_duration = item_duration[['session_id', 'item_id', 'item_session_duration']]
  item_duration = item_duration[['session_id', 'item_id', 'item_duration']]
  item_session_duration = item_session_duration.fillna(0)
  item_session_druation = item_session_duration.reset_index(drop=True)
  return item_session_duration

In [0]:
item_session_duration = get_item_session_duration(item_duration, session_duration)
item_session_duration[item_session_duration.item_session_duration>0]

session_id  item_id  item_session_duration
25        3599a6f709eab  2795374               1.000000
65        ec139e10b9238  1032816               0.550489
79        325fafb5fa450    65685               0.256917
126       7157899be2839  2552514               1.000000
185       4c6062d7cefe4   138930               0.080645
...                 ...      ...                    ...
15373250  92b7ab1287edf   207081               0.009609
15373251  92b7ab1287edf    50584               0.056280
15373273  6b66cb0cfb518  3811156               0.230769
15373274  6b66cb0cfb518  2337580               0.769231
15373323  62728015bec05  6617798               0.036778

[530222 rows x 3 columns]

#### item_interactions

In [0]:
def get_item_interactions(data, item_id):
  item_interactions = data.groupby(['session_id', 'reference']).step.count().to_frame().reset_index()
  item_interactions = item_interactions.rename(columns={'reference':'item_id', 'step':'item_interactions'})
  item_interactions = item_id.merge(item_interactions, left_on=['session_id', 'item_id'], right_on=['session_id', 'item_id'], how='left')
  item_interactions = item_interactions.fillna(0)
  item_interactions = item_interactions.reset_index(drop=True)
  return item_interactions

In [0]:
item_interactions = get_item_interactions(train, item_id)
item_interactions[item_interactions.item_interactions>0]

session_id  item_id  item_interactions
5         aff3928535f48  1257342                1.0
25        3599a6f709eab  2795374               35.0
65        ec139e10b9238  1032816                4.0
79        325fafb5fa450    65685               38.0
82        325fafb5fa450  1320460                1.0
...                 ...      ...                ...
15373274  6b66cb0cfb518  2337580                8.0
15373318  e7916050980d9  8985292                1.0
15373323  62728015bec05  6617798               15.0
15373326  62728015bec05  1161323                1.0
15373343  62728015bec05  2712342                1.0

[1168107 rows x 3 columns]

#### maximum_step

In [0]:
def get_maximum_step(data, item_id):
  maximum_step = data.groupby('session_id', sort=False).step.max().to_frame().reset_index()
  maximum_step = maximum_step.rename(columns={'step':'maximum_step'})
  maximum_step = item_id.merge(maximum_step, on='session_id', how='left')
  maximum_step = maximum_step.reset_index(drop=True)
  return maximum_step

In [0]:
maximum_step = get_maximum_step(train, item_id)
maximum_step

session_id  item_id  maximum_step
0         aff3928535f48    55109            16
1         aff3928535f48   129343            16
2         aff3928535f48    54824            16
3         aff3928535f48  2297972            16
4         aff3928535f48   109014            16
...                 ...      ...           ...
15373343  62728015bec05  2712342            19
15373344  62728015bec05    48497            19
15373345  62728015bec05    11933            19
15373346  62728015bec05  1714483            19
15373347  62728015bec05  1236687            19

[15373348 rows x 3 columns]

#### top_list

In [0]:
def get_top_list(item_rank):
  top_list = item_rank[['session_id', 'item_rank']]
  top_list['top_list'] = top_list.apply(lambda x: 1 if x['item_rank'] < 6 else 0, axis=1)
  top_list = top_list.reset_index(drop=True)
  return top_list

In [0]:
top_list = get_top_list(item_rank)
top_list

session_id  item_rank  top_list
0         aff3928535f48          1         1
1         aff3928535f48          2         1
2         aff3928535f48          3         1
3         aff3928535f48          4         1
4         aff3928535f48          5         1
...                 ...        ...       ...
15373343  62728015bec05         21         0
15373344  62728015bec05         22         0
15373345  62728015bec05         23         0
15373346  62728015bec05         24         0
15373347  62728015bec05         25         0

[15373348 rows x 3 columns]

In [0]:
dataframes = [item_id, price, item_rank, price_rank, clickout, session_duration, item_duration, item_session_duration, item_interactions, maximum_step, top_list]
local_data = pd.concat(dataframes)
local_data

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


## Merging Global Features to Local Feautres

In [0]:
GlobalPath = './Datasets/clean_data/ItemsFeatures/item_global.csv'
LocalPath = './Datasets/clean_data/ItemsFeatures/item_local.csv'

GlobalData = pd.read_csv(GlobalPath)
LocalData = pd.read_csv(LocalPath)

TrainData = LocalData.merge(GlobalData, on='item_id', how='left')
TrainData

In [0]:
TrainData[TrainData.MeanRank.isnull()]

In [0]:
TrainData.item_id.nunique(), item_metadata.item_id.nunique()

In [0]:
len(set(TrainData.item_id.unique()) - set(item_metadata.item_id.unique()))

706

Although item_metadata should have all the items in the training set, 706 of the items don't exits in item_metadata dataset. NaN values will be filled with mean value of each attribute.

In [0]:
TrainData.drop(columns=['Unnamed: 0', 'properties'], inplace=True)
NaNcolumns = ['NumberOfProperties', 'NumberInImpressions', 'NumberInReferences', 'NumberAsClickout', 'NumberAsFinalClickout',
              'FClickoutToImpressions', 'FClickoutToReferences', 'FClickoutToClickout', 'MeanPrice', 'AveragePriceRank']
for column in NaNcolumns:
  MeanValue = TrainData[column].mean()
  TrainData[column] =  TrainData[column].fillna(MeanValue)

Empty DataFrame
Columns: [session_id, item_id, price, item_rank, price_rank, clickout, session_duration, item_duration, item_session_duration, item_interactions, maximum_step, top_list, NumberOfProperties, NumberInImpressions, NumberInReferences, NumberAsClickout, NumberAsFinalClickout, FClickoutToImpressions, FClickoutToReferences, FClickoutToClickout, MeanPrice, AveragePriceRank]
Index: []

All above functions in the above cells are defined in one function in a class in order to make the code reusable for validation and test sets. The class is called data_transformation, and the main function in the class is transform_data(data), and takes only the Pandas Dataframe as an input, and gives the ready data for standardizing and then processing.

## Applying transform_data Function

All the functions are saved in a class that can be imported and apply the function in just one line.

In [0]:
from data_transformation import data_transformation

TrainData = data_transformation.transform_data(train)
TrainData

/content/drive/My Drive/Trivago/Project/TrivagoRecommenderSystem/data_transformation.py:17: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  item_id['impressions'] = item_id['impressions'].apply(lambda x: x.split('|'))
/content/drive/My Drive/Trivago/Project/TrivagoRecommenderSystem/data_transformation.py:35: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  price['prices'] = price['prices'].apply(lambda x: x.split('|'))
/content/drive/My Drive/Trivago/Project/TrivagoRecommenderSystem/data_transformation.py:44: Sett

session_id  item_id  price  ... MaxPrice MinPrice  AveragePriceRank
0         aff3928535f48    55109    162  ...   1905.0     65.0         11.840753
1         aff3928535f48   129343     25  ...    388.0     18.0         10.166667
2         aff3928535f48    54824    150  ...    408.0     42.0         11.291196
3         aff3928535f48  2297972    143  ...    183.0    101.0          9.647059
4         aff3928535f48   109014    101  ...    628.0     70.0         11.262887
...                 ...      ...    ...  ...      ...      ...               ...
12586682  62728015bec05  2712342     73  ...    881.0     54.0         12.950000
12586683  62728015bec05    48497    169  ...    253.0     99.0         11.940594
12586684  62728015bec05    11933     87  ...    311.0     69.0         12.262411
12586685  62728015bec05  1714483    485  ...    586.0    141.0         12.510204
12586686  62728015bec05  1236687    171  ...    500.0    113.0         12.191667

[12586687 rows x 25 columns]

In [0]:
TrainData.to_csv('./Datasets/clean_data/Sets/train.csv', index=False)

Merging properties to the TrainData as it carries valuable information such as rating, but at the same time because the dataset will be huge and session crashes, further step needs to be taken.

In [0]:
# TrainData.item_id = TrainData.item_id.apply(lambda x: str(x))
# TrainData.merge(properties_encodedDF, on='item_id', how='left')